[Feature] Add WidowXAI Robot, specify robot ids in example scripts

hu-po opened the pull request:
Adding the new [WidowXAI robot](https://www.trossenrobotics.com/widowx-ai) from [Trossen Robotics](https://github.com/TrossenRobotics). Implementation follows the format of the existing WidowX250S robot.

Added `-robot_id` and `-r` args to specify robot for environments in example scripts.

Robot assets are located [here](https://github.com/hu-po/ManiSkill-WidowXAI), again following the format of the [widowx250s](https://github.com/haosulab/ManiSkill-WidowX250S)

@StoneT2000 do you have a command for the png at `ManiSkill/docs/source/robots/images`?

StoneT2000 commented on mani_skill/examples/benchmarking/gpu_sim.py:
```diff
@@ -20,6 +20,8 @@
 @dataclass
 class Args:
     env_id: Annotated[str, tyro.conf.arg(aliases=["-e"])] = "PickCube-v1"
+    robot_uids: Annotated[str, tyro.conf.arg(aliases=["-r"])] = "widowxai"
```
make the default None so the task uses the default robot of the task.

StoneT2000 commented on mani_skill/examples/benchmarking/gpu_sim.py:
```diff
@@ -20,6 +20,8 @@
 class Args:
     env_id: Annotated[str, tyro.conf.arg(aliases=["-e"])] = "PickCube-v1"
+    robot_uids: Annotated[str, tyro.conf.arg(aliases=["-r"])] = "widowxai"
+    """Some environments support multiple robots"""
```
maybe make this more informative like

"Pass in a robot uid to change the robot loaded into the environment instead of using the default one"

StoneT2000 commented on examples/baselines/ppo/ppo_fast.py:
```diff
@@ -86,6 +86,8 @@ class Args:
     """frequency to save training videos in terms of iterations"""
     control_mode: Optional[str] = "pd_joint_delta_pos"
     """the control mode to use for the environment"""
+    robot_uids: Optional[str] = None
```
thanks for modifying and testing the PPO code! 

do you know if tyro cli supports generic kwargs to pass via cli? I guess there are a lot of env kwargs like control mode robot uids etc that one can pass but I would prefer not to have a line of code for each.

hu-po replied:
np, ty for maintaining this repo, legit one of the best sim stacks for manip

dunno, there probably is some way to do what you are describing, it is definitely ugly/verbose to just keep appending kwargs. BUT for the sake of this PR I think its better to keep it minimal and separately create some kind of github issue/ticket thing for the cleaner kwargs

StoneT2000 commented on docs/source/robots/arms/index.md:
```diff
@@ -31,4 +31,17 @@ Robot UID: `widowx250s`
\ No newline at end of file
+Robot Description: [https://github.com/haosulab/ManiSkill-WidowX250S](https://github.com/haosulab/ManiSkill-WidowX250S)
+
+## Widow X AI
```
thanks for documenting!

Meanwhile need to put up an issue to auto-generate the robot docs like the task docs in the future to make things easier to maintain.

hu-po replied:
:memo:

I feel like github is probably working towards some kind of github action agent that does exactly this

hu-po commented:
https://github.com/user-attachments/assets/3e3f9674-ecc3-4ac1-972c-59d7f4483b58

I was hoping for a clear video of picking up the cube to close out this PR, but despite sweeping over max_steps and seeds all I got was some lucky solves that are really more of a push than a pick. I think the biggest issue is that this robot is small so it starts far away and the gripper is barely big enough to grasp the cube. Collisions look good, but the cube needs to be smaller and nearer to the robot. Modifying the task with even more kwargs feels like it will just add more verbosity and cruft. Imo this feels okay enough to just ship, but I understand if yall would rather not include something unless it passes convincingly.

StoneT2000 commented:
Its strange it can't learn to grasp an object. We generally do want some sign of life that manipulation works before adding a manipulation system.

A couple suggestions:
- Initialize the robot so that the gripper is in a similar pose as what the panda starts off in. This means less steps are needed to solve the task (and you can keep the max_steps=50)
- Your PPO args could probably be optimized. The examples.sh file should be removed in favor of the baselines.sh file in there. The baselines.sh ones are better tuned and you can see the example hyperparams used for pickcube with the default robot (panda)
- You should double check if the is_grasping function is working as intended. It depends on the orientation of the gripper links and some robots have this orientation flipped compared to others. A way to check is to create the pick cube env, use pd_joint_pos action space, and script the joint positions so  that the robot moves to the cube, then tries to grasp it, and then check the returned `info` object for whether the env thinks the robot is grasping it.

```
env = gym.make("PickCube-v1", robot_uids=..., control_mode="pd_joint_pos")
env.reset(seed=0)
joint_pos_points = [...]
for joint_pos in joint_pos_points:
    for _ in range(50):
        obs, rew, term, trunc, info = env.step(joint_pos)
        env.render_human()
        print(info)
```
to get these joint pos reference points, you can open the viewer, and click the end-effector and click the transform tab to drag the robot to the cube and figure out where it is.

hu-po commented:
Quick update on this project. **What has been done:**

- Created a new `ready_to_grasp` keyframe for the robot, this starts it closer to the goal and reduces the complexity of the task.
- Removed `examples.sh` in favor of `baselines.sh`
- Added a black material to the robot meshes
- Verified grasping function by stepping through the environment:
![is_grasping](https://github.com/user-attachments/assets/22a5c6a2-f60c-4b16-ab70-049a0c711235)
- Copied over the control_mode and friction material from the `panda` robot.
- Sweeped over `control_mode`, `seed`, and `num_steps` for the `PickCube-v1` task. [wandb logs](https://api.wandb.ai/links/hug/cetio32z). Here are some cherry picked results:

https://github.com/user-attachments/assets/cd497676-b41d-4e10-8c0a-79d9ee5d8c44


https://github.com/user-attachments/assets/1ad72d72-c732-4179-a04d-8afa68d91b36


https://github.com/user-attachments/assets/7af2eb95-9806-4ccc-96c3-02025c1bab57


https://github.com/user-attachments/assets/bf0b3f9a-c03f-4e0b-ba95-5879e6dc8b71


- From visual inspection, the collisions and interactions between the robot and the cube are definitely correct. The different control modes have their phenotypical quirks, but none of them work drastically better or worse than the others. The longer episode length also doesn't really help that much either. It still feels like the task is just too difficult.

**What will be done:** I will run a similar sweep but on the `PushCube-v1` task. I think with this easier task we should be able to get much better results.

StoneT2000 commented:
for pick cube another consideration is the robot is too small anyway, its reach is not as large as the franka / panda arm and this task has done its randomization with the panda arm in mind.

Try lowering the amount of cube pose randomization + goal site randomization (eg the goal should probably not be so high since this arm can't reach it)

moreover what is the ppo script you use at the moment?

hu-po commented:
I am using `ppo_fast.py`, which is what is used in `baselines.sh`. I also tried `ppo.py` and `ppo_rgb.py`, both run clean but don't learn as fast. [My sweeping scripts](https://github.com/haosulab/ManiSkill/blob/757123c401e9795aa4fe668ddd4b1a353ac8311a/examples/baselines/ppo/test-widowxai-pickcube.sh) are in this branch right now, but ill remove them before the merge.

StoneT2000 commented:
oh definitely do not need to sweep so many control modes. It should work with pd_joint_delta_pos. If not then something is modelled poorly

Also do not use any absolute control spaces (like pd_joint_pos or pd_ee_pose), RL does not learn from those easily unless you modify the reward function more to penalize large actions.

hu-po commented:
`PushCube-v1` works much better, it still has reach issues but it adorably discovers cube curling!

https://github.com/user-attachments/assets/5a38381e-a542-463d-a6c1-01da73ca13d8

https://github.com/user-attachments/assets/7fc3424b-b114-465e-bb3e-32d6c8124d4f

@StoneT2000 I removed the scaffolding/test code so this code should be ready to merge. I am stepping away from this work for now: I am receiving the real robot soon and will focus on that the next couple weeks (e.g. testing out lerobot's imitation learning pipeline on the real robot). After that I will come back to this ManiSkill project and work towards the original goal of some kind of real2sim2real. We can just merge this now and then I can start a new pr later, or perhaps you would prefer one larger PR over smaller PRs. Up to you as the maintainer.

StoneT2000 commented:
I will probably not merge for now given that it can't seem to pick up the cube yet (I haven't looked in depth enough into the robot modeling). I did notice the gripper friction is 4.0, try setting it to 2.0 instead? Training should probably take no more than 5 minutes on a 4090 with ppo_fast. You should only try using the `pd_joint_delta_pos` action space

Another thing to try is to make the cube a bit smaller just for sanity check. It could be relatively too big for this robot.

lukeschmitt-tr commented:
Hello @hu-po @StoneT2000 

I work at Trossen Robotics. It's really great to see this effort and we would love to contribute! Is there anything we can do to help get this over the finish line and officially added to ManiSkill?

Please note that we have [fixed an error](https://github.com/TrossenRobotics/trossen_arm_description/pull/5) in the description file somewhat recently and wanted to make sure that you were using [the latest version of the wxai_follower URDF](https://github.com/TrossenRobotics/trossen_arm_description/blob/2624c550ffd893f3a2fb63a6592d80d51cd4d983/urdf/generated/wxai/wxai_follower.urdf).

StoneT2000 commented:
I should have more time next week. Seeing a Trossen employee interested merging this in is good motivation! I can try and dedicate some time next week to help @hu-po finalize this merge and debug the robot modelling.

hu-po commented:
That error is interesting, last I worked on this I was just tuning numbers but wasn't getting any luck with ol' PPO. @lukeschmitt-tr we should replace the path to the zip here `mani_skill/utils/assets/data.py` with something that Trossen officially maintains.

@StoneT2000 thanks as always for the help and attention. I digged up some old videos, my suspicion is something to do with the friction on the fingertips? I tried a couple cube sizes and initial starting positions, but the most they do is just push the cube around.

https://github.com/user-attachments/assets/73ac137d-b059-47d5-8632-6159d13ef93f


https://github.com/user-attachments/assets/59945eab-3bc3-4614-8982-66e9c82e15c3

StoneT2000 commented:
@lukeschmitt-tr which widowxai robot model is the updated URDF for? The real hardware specifically

@hu-po which model of the robot from Trossen are you using exactly? Is it different?

The one we have modeled now in simulation (the ones in the videos) is lightly modeled by us previously for the SIMPLER real2sim eval project, based on the hardware used in the Bridge-v2 Dataset.

StoneT2000 commented:
I also noticed @lukeschmitt-tr there are a few other robots provided now by Trossen including a cool dual arm mobile manipulator.  Is the mobile_ai robot stable now and the model is more or less complete? If so I can help model it, we don't have a decent dual arm mobile robot modelled at the moment in simulation, just single arm (fetch only actually which is discontinued).

StoneT2000 commented:
One more thing I observed in the URDF provided @lukeschmitt-tr is that there isn't a mimic joint tag for the left and right carriage joints. Is this expected?

StoneT2000 commented:
[widowxai.zip](https://github.com/user-attachments/files/20619348/widowxai.zip)

These are the updated URDF and SRDF models

@hu-po to test the working setup now, you can checkout my branch which is based on your PR `widowxai-rl`

And then in the PPO examples folder run the baseline code for this robot:

```
seeds=(9351 4796 1788)
for seed in ${seeds[@]}
do
  python ppo_fast.py --env_id="PickCubeWidowXAI-v1" --seed=${seed} \
    --num_envs=4096 --num-steps=8 --update_epochs=8 --num_minibatches=32 \
    --total_timesteps=50_000_000 \
    --num_eval_envs=16 \
    --cudagraphs --exp-name="ppo-PickCubeWidowXAI-v1-state-${seed}-walltime_efficient" \
    --track
done
```

I made the task slightly harder (its quite easy to solve) by making cube slightly bigger and goal height randomization larger.

StoneT2000 commented:
This PR

Adds two new robots
- WidowXAI
- WidowXAI follower (the above but with wrist camera mount)

Some TODOs is the wrist camera mount has not been calibrated in simulation to match the real world. Quality rating I keep both at B since I have not done system ID on them or calibrated the simulation camera intrinsics + extrinsics. Another is we need the final link for where to host assets

The PR also
- modifies the robot documentation generation script to accept a single robot id to update instead of updating all at once. It seems image files have small differences when written to disk by cv2 depending on machine which easily causes file differences picked up by git
- updates the doc strings for PickCube (the baseline benchmark environment for testing if a robot can manipulate) to be formatted and adds missing docs for the SO100 robot as well
- remove support for passing robot uid to the ppo training script. It can easily confuse users into thinking RL still works if you swap a robot. Most times it doesn't as most environments are designed with reward functions and settings for one robot embodiment in mind

StoneT2000 commented:
Thank you for the fast replies @hu-po and @lukeschmitt-tr

lukeschmitt-tr commented:
Thanks to @hu-po for getting this effort started and to @StoneT2000 for the reviews, advice, and merge!

> Quality rating I keep both at B

@StoneT2000 What would it take for a better rating?

StoneT2000 commented:
Testing on a real robot!

Right now there hasn't been any system ID done. For the wrist camera we need the intrinsic matrix and exact positioning. Preferably if we can test on more than one of the same to see if there is any small hardware differences between robots

For system ID essentially we need to tune the stiffness and damping parameters for the joint position controller to ensure the sim follows the real robot behavior.

B just means it simulates fine but not tuned. Following the mujoco menagerie classification approach.
